### PR TITLE
allow FitFactories (custom localization routines) from plugins 

### PR DESCRIPTION
**Pull request by barentine**
Addresses issue on our roadmap (official or not, I'm not sure, but certainly discussed!)

**Is this a bugfix or an enhancement?**
enhancement
**Proposed changes:**
- allow fitfactories to be plugged using the new yaml plugin registration (NOTE - need to update `pyme-plugin`)
- small update to include reports folder in old-style plugin reg.

**Checklist:**
The below is a list of things what will be considered when reviewing PRs. It is not prescriptive, and does not
imply that PRs which touch any of these will be rejected but gives a rough indication of where there is a potential 
for hangups (i.e. factors which could turn a 5 min review into a half hour or longer and shunt it to the bottom
of the TODO list).

- [x] Does the PR avoid variable renaming in existing code, whitespace changes, and other forms of tidying? [There is a place for code tidying, but it makes reviewing 
much simpler if this is kept separate from functional changes. The auto-formatting performed by some editors is particulaly egregious and can lead to files with thousands
of non-functional changes with a few functional changes scattered amoungst them]

Note that I changed the fit factory import to use import_module, which the docs say to use instead of __import__ these days

If an enhancement (or non-trivial bugfix):

- [x] Has this been discussed in advance (feature request, PR proposal, email, or direct conversation)?
- [x] Does this change how users interact with the software? How will these changes be communicated?
We really ought to update the docs accordingly
- [x] Does this maintain backwards compatibility with old data?
- [x] Does this change the required dependencies?
- [x] Are there any other side effects of the change?

### CODE REVIEW

**Comment by barentine** [comment]
This should now allow most fit factory imports to work for plugin fitfactories, with the exception of imports that are more specific, e.g. `self.interpolator = __import__('PYME.localization.FitFactories.Interpolators.' + interpModule , fromlist=['PYME', 'localization', 'FitFactories', 'Interpolators']).interpolator`

For full functionality throughout PYME we would have to do a bit more fit factory metadata collection. But I'd be happy for now with what's here, so I'm going to ~stop.

**Comment by barentine** [comment]
Some notes on documentation I think we'll want to include:
- fithelpers need doc, including e.g. that first derivatives if provided need to be raveled
- Need to tell people about fitresult key name oddities, like x0 and y0 being special (called mine x and y the first time), and underscores being problematic

**Comment by barentine** [comment]
Any thoughts on this @David-Baddeley ?

File: PYME/localization/FitFactories/__init__.py [hunk 1]
```diff
@@ -22,8 +22,13 @@
 ################
 import glob
 import os
+from PYME import config
+import importlib
 
-__all__ = [os.path.splitext(os.path.split(p)[-1])[0] for p in glob.glob(__path__[0] + '/[a-zA-Z]*.py')]
+# discover PYME-internal fitfactories and fully qualify them
+__all__ = ['PYME.localization.FitFactories.' + os.path.splitext(os.path.split(p)[-1])[0] for p in glob.glob(__path__[0] + '/[a-zA-Z]*.py')]
```
Comment (David-Baddeley) [inline PYME/localization/FitFactories/__init__.py]: `__all__` is a special variable (although this looks like it's my mistake, just propagated)

File: PYME/localization/FitFactories/__init__.py [hunk 2]
```diff
@@ -52,3 +61,20 @@
...
+    ----------
+    fit_factory : str
+        fit factory module name, e.g. LatGaussFitFR
+
+    Returns
+    -------
+    module
+        loaded fit factory module
+    """
+    pkg = package[fit_factory]
```
Comment (David-Baddeley) [inline PYME/localization/FitFactories/__init__.py]: Leaves the possibility for name collisions in plugins - maybe better to:

simple name -> PYME internal (check first)
fully resolved (i.e. `"my_plugin.my_factory"`) -> from plugin

File: PYME/localization/remFitBuf.py [hunk 3]
```diff
@@ -346,7 +348,8 @@ def __init__(self, dataSourceID, frameIndex, metadata, dataSourceModule=None, re
     @property
     def fitMod(self):
         if self._fitMod is None:
-            self._fitMod = __import__('PYME.localization.FitFactories.' + self.fitModule, fromlist=['PYME', 'localization', 'FitFactories']) #import our fitting module
+            pkg = PYME.localization.FitFactories.package[self.fitModule]
+            self._fitMod = importlib.import_module('.'.join([pkg, self.fitModule]), pkg)  # import our fitting module
```
Comment (David-Baddeley) [inline PYME/localization/remFitBuf.py]: is there any reason not to use `import_fit_factory()` here?

**Comment by David-Baddeley** [comment]
Sorry I've taken so long to have a look at this ... No major issues, but:

- I'd like to use fully resolved names (rather than just shortened names) for plugin modules as it a) makes it clear that it's coming from a plugin (which is good for reproducibility, and directing support queries to the right place) and b) reduces the chance of collisions .
- I'll probably change the `__all__` think to something non-broken.

**Comment by d-perez-1** [comment]
image
- AESB & DDP working a fix on opening PYMEImage

```
Traceback (most recent call last):
  File "C:\ProgramData\Miniconda3\lib\site-packages\wx\core.py", line 3259, in <lambda>
    lambda event: event.callable(*event.args, **event.kw) )
  File "c:\userfiles\code\python-microscopy\PYME\DSView\dsviewer.py", line 470, in LoadData
    vframe = DSViewFrame(im, None, im.filename, mode = mode)
  File "c:\userfiles\code\python-microscopy\PYME\DSView\dsviewer.py", line 181, in __init__
    self.CreateFoldPanel()
  File "c:\userfiles\code\python-microscopy\PYME\ui\AUIFrame.py", line 139, in CreateFoldPanel
    genFcn(self.sidePanel)
  File "c:\userfiles\code\python-microscopy\PYME\DSView\modules\LMAnalysis.py", line 258, in GenAnalysisPanel
    self._populateCustomAnalysisPanel(self.customOptionsPan, self.customOptionsSizer)
  File "c:\userfiles\code\python-microscopy\PYME\DSView\modules\LMAnalysis.py", line 171, in _populateCustomAnalysisPanel
    pkg = PYME.localization.FitFactories.package[fitMod]
KeyError: 'LatGaussFitFR'
```

**Pull request merged**